webui/react/src/hooks/useSettings.ts

```diff
@@ -1,359 +1,294 @@
-import { ErrorType } from '../shared/utils/error';
-
-import useStorage from './useStorage';
-
-export const BaseType = {
-  Boolean: 'Boolean',
-  Float: 'Float',
-  Integer: 'Integer',
-  Object: 'Object',
-  String: 'String',
-} as const;
-
-export type BaseType = ValueOf<typeof BaseType>;
-
-const PathChangeType = {
-  None: 'none',
-  Push: 'push',
-  Replace: 'replace',
-} as const;
-
-type PathChangeType = ValueOf<typeof PathChangeType>;
-
-type GenericSettingsType = Primitive | Primitive[] | Record<number, number[]> | undefined;
-type GenericSettings = Record<string, GenericSettingsType>;
-type PathChange<T> = { querySettings: Partial<T>; type: PathChangeType };
-
-/*
- * defaultValue     - Optional default value. `undefined` as ultimate default.
- * skipUrlEncoding  - Avoid preserving setting in the URL query param.
- * storageKey       - If provided, save/load setting into/from storage.
- * type.baseType    - How to decode the string-based query param.
- * type.isArray     - List based query params can be non-array.
- */
-export interface SettingsConfigProp {
-  defaultValue?: GenericSettingsType;
-  key: string;
+import { Settings, SettingsProvider, UserSettings } from './useSettingsProvider';
+
+export interface SettingsConfigProp<A> {
+  defaultValue: A;
   skipUrlEncoding?: boolean;
-  storageKey?: string;
-  type: {
-    baseType: BaseType;
-    isArray?: boolean;
-  };
+  storageKey: string;
+  type: t.Type<A>;
 }
 
-export interface SettingsConfig {
-  applicableRoutespace?: string;
-  settings: SettingsConfigProp[];
+export interface SettingsConfig<T> {
+  applicableRoutespace: string;
+  settings: { [K in keyof T]: SettingsConfigProp<T[K]> };
   storagePath: string;
 }
 
-/*
- * Provide the ability to override hook options with
- * dynamic values during initialization.
- */
-export interface SettingsHookOptions {
-  storagePath?: string;
+interface UserSettingUpdate extends UpdateUserSettingParams {
+  userId: number;
 }
 
-export type UpdateSettings<T> = (newSettings: Partial<T>, push?: boolean) => void;
+export type UpdateSettings = (updates: Settings, shouldPush?: boolean) => Promise<void>;
+export type ResetSettings = (settings?: string[]) => void;
 
-export interface SettingsHook<T> {
+export type UseSettingsReturn<T> = {
   activeSettings: (keys?: string[]) => string[];
-  resetSettings: (keys?: string[]) => void;
-  settings: T;
-  updateSettings: UpdateSettings<T>;
-}
-
-export const validateBaseType = (type: BaseType, value: unknown): boolean => {
-  if (type === BaseType.Boolean && isBoolean(value)) return true;
-  if (type === BaseType.Float && isNumber(value)) return true;
-  if (type === BaseType.Integer && isNumber(value) && Math.ceil(value) === Math.floor(value))
-    return true;
-  if (type === BaseType.String && isString(value)) return true;
-  if (type === BaseType.Object && isObject(value)) return true;
-  return false;
+  isLoading: boolean;
+  resetSettings: ResetSettings;
+  settings: T | null;
```

Comment:
+1

Reply:
it might lead to a bit of janky loading and maybe we want to transition it to a `Loadable` pattern to avoid that in the future, but just for this initial PR i think it's less risky to keep the API consistent

Reply:
idk if its applicable here, but we can use `suspense` when useSettings is loading data

Reply:
Can you provide an example of how that would work?

Reply:
https://beta.reactjs.org/apis/react/Suspense

Reply:
> Can you provide an example of how that would work?

we can just have the whole app in a loading state until the initial fetch?

> You don't actually want to "fall back" because that will cause an incorrect state

that's what I meant by "janky loading." i agree that it's not what we want but imo is preferable to a page crash. i think bringing in loadable would be good, but this particular PR is big as is.

Reply:
@ClaireNeveu I think our previous message might have crossed paths in flight

Reply:
> https://beta.reactjs.org/apis/react/Suspense

@keita-determined  thanks for linking, need to freshen up on suspense on a bit. can you elaborate on what in particular you had in mind though?

Reply:
Unless I'm missing something, suspense only works with React.lazy which isn't applicable here.

Reply:
`Suspense` works with api call. when API is not resolved yet, `Suspense` will fall back. 
When API is resolved `suspense` will show the child components. 
The link i sent doesnt fully covered all cases
prob this one is better https://dev.to/darkmavis1980/a-practical-example-of-suspense-in-react-18-3lln